### PR TITLE
Feature/algo eval

### PR DESCRIPTION
**Pull request by maxhuettenrauch**
# Changes

## Dependencies

- New extra "eval"

## Api Extension
- `Experiment` and  `ExperimentConfig` now have a `name`, that can however be overridden when `Experiment.run()` is called
- When building an `Experiment` from an `ExperimentConfig`, the user has the option to add info about seeds to the name.
- New method in `ExperimentConfig` called `build_default_seeded_experiments`
- `SamplingConfig` has an explicit training seed, `test_seed` is inferred.
- New `evaluation` package for repeating the same experiment with multiple seeds and aggregating the results (important extension!). Currently in alpha state.
- Loggers can now restore the logged data into python by using the new `restore_logged_data`

## Breaking Changes
- `AtariEnvFactory` (in examples) now receives explicit train and test seeds
- `EnvFactoryRegistered` now requires an explicit `test_seed`
- `BaseLogger.prepare_dict_for_logging` is now abstract

### CODE REVIEW

**Comment by maxhuettenrauch** [comment]
@MischaPanch @bordeauxred please have a look

File: examples/mujoco/mujoco_ppo_hl_multi.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                else None,
+                dist_fn=DistributionFunctionFactoryIndependentGaussians(),
+            ),
+        )
+        .with_actor_factory_default(hidden_sizes, torch.nn.Tanh, continuous_unbounded=True)
+        .with_critic_factory_default(hidden_sizes, torch.nn.Tanh)
+        .build_default_seeded_experiments(num_experiments)
+    )
+
+    for experiment_name, experiment in experiments.items():
```
Comment (MischaPanch) [inline examples/mujoco/mujoco_ppo_hl_multi.py]: Let's add a TODO here to parallelize this asap. In fact, already here something like a hydra launcher could be useful, but maybe we do that in a separate PR later

File: examples/mujoco/mujoco_ppo_hl_multi.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+
+def eval_experiments(log_dir: str):
+    results = RLiableExperimentResult.load_from_disk(log_dir, "PPO", None)
+    eval_results(results)
+
+
+if __name__ == "__main__":
+    # logging.run_cli(main)
+    # experiment_config = ExperimentConfig(watch=False)
```
Comment (MischaPanch) [inline examples/mujoco/mujoco_ppo_hl_multi.py]: main should be executed, right?

File: examples/mujoco/tools.py [hunk 3]
```diff
@@ -5,11 +5,111 @@
...
+@dataclass
+class RLiableExperimentResult:
+    exp_dir: str
+    algorithms: list[str]
+    score_dict: dict[str, np.ndarray]  # (n_runs x n_epochs + 1)
+    env_steps: np.ndarray  # (n_epochs + 1)
+    score_thresholds: np.ndarray
+
+    @staticmethod
+    def load_from_disk(exp_dir: str, algo_name: str, score_thresholds: np.ndarray | None):
```
Comment (MischaPanch) [inline examples/mujoco/tools.py]: I feel like loading should only require one path parameter. Is it possible to have default behavior without passing algo_name and trhesholds? Should the latter be None by default?

File: examples/mujoco/tools.py [hunk 4]
```diff
@@ -5,11 +5,111 @@
...
 
+from tianshou.highlevel.experiment import Experiment
+
+
+@dataclass
+class RLiableExperimentResult:
+    exp_dir: str
+    algorithms: list[str]
+    score_dict: dict[str, np.ndarray]  # (n_runs x n_epochs + 1)
+    env_steps: np.ndarray  # (n_epochs + 1)
```
Comment (MischaPanch) [inline examples/mujoco/tools.py]: in tianshou 1 epoch has many steps, is it not true in this context?

File: examples/mujoco/tools.py [hunk 5]
```diff
@@ -5,11 +5,111 @@
...
+            score_thresholds=score_thresholds,
+            exp_dir=exp_dir,
+        )
+
+
+def eval_results(results: RLiableExperimentResult):
+    import matplotlib.pyplot as plt
+    import scipy.stats as sst
+    import seaborn as sns
+    from rliable import library as rly
```
Comment (MischaPanch) [inline examples/mujoco/tools.py]: rliable should become an extra dependency in the pyproject. On import error we should tell the user to install with `-E rliable`

As to the other dependencies - are they not part of tianshou yet? Do we need the local import? If they aren't, the extra should be something like `eval` and contain all these dependencies

Comment (MischaPanch) [reply examples/mujoco/tools.py]: Alternatively, if theyre not too big, we could just add them to the direct dependencies. Up to rliable, they're very standard. If rliable itself is not heavyweight, we can just add it, I think

File: examples/mujoco/tools.py [hunk 6]
```diff
@@ -5,11 +5,111 @@
...
+    from rliable import plot_utils
+
+    iqm = lambda scores: sst.trim_mean(scores, proportiontocut=0.25, axis=0)
```
Comment (MischaPanch) [inline examples/mujoco/tools.py]: Why the hardcoded 0.25 and reps=50000. Would the user never be interested in adjusting them?

File: examples/mujoco/tools.py [hunk 7]
```diff
@@ -5,11 +5,111 @@
...
+    plot_utils.plot_sample_efficiency_curve(
+        results.env_steps,
+        iqm_scores,
+        iqm_cis,
+        algorithms=results.algorithms,
+        xlabel=r"Number of env steps",
+        ylabel="IQM episode return",
+        ax=ax,
+    )
+    plt.savefig(os.path.join(results.exp_dir, "iqm_sample_efficiency_curve.png"))
```
Comment (MischaPanch) [inline examples/mujoco/tools.py]: I feel like the path should be configurable with a default value.

The function needs a docstring explaining that images will be saved to disk, I wouldn't have expected it. Is it possible to easily separate plotting from saving here? I feel like saving itself should be optional

Comment (MischaPanch) [reply examples/mujoco/tools.py]: Note that the user might want to decide how to save files - as svg, with different qualities, to different places, etc. You could just return figure or ax objects and let the downstream code save them

File: tianshou/highlevel/experiment.py [hunk 8]
```diff
@@ -86,6 +87,31 @@
 log = logging.getLogger(__name__)
 
 
+def shortener(input_string: str | None = None, length: int = 1) -> str:
+    """Shorten the input string by keeping only the first `length` characters of each word.
+
+    If the input string is None or empty, return "default".
```
Comment (MischaPanch) [inline tianshou/highlevel/experiment.py]: No need for such a behavior in this function

File: tianshou/highlevel/experiment.py [hunk 9]
```diff
@@ -86,6 +87,31 @@
 log = logging.getLogger(__name__)
 
 
+def shortener(input_string: str | None = None, length: int = 1) -> str:
+    """Shorten the input string by keeping only the first `length` characters of each word.
```
Comment (MischaPanch) [inline tianshou/highlevel/experiment.py]: How are words separated? The separator should be configurable

File: tianshou/highlevel/experiment.py [hunk 10]
```diff
@@ -86,6 +87,31 @@
...
+
+    If the input string is None or empty, return "default".
+    """
+    if input_string is None or len(input_string) == 0:
+        return "default"
+    output_parts = []
+
+    for part in input_string.split(","):
+        key, value = part.split("=")
```
Comment (MischaPanch) [inline tianshou/highlevel/experiment.py]: This doesn't do what the docstring says, nowhere is `=` mentioned.

What is the reason behind creating this function at all? Let's discuss in a short call instead of here

File: tianshou/highlevel/experiment.py [hunk 11]
```diff
@@ -442,6 +484,46 @@ def build(self) -> Experiment:
...
+            exp = self.build()
+
+            full_name = ",".join(
+                [
+                    f"experiment_seed={exp.config.seed}",
+                    f"train_seed={exp.sampling_config.train_seed}",
+                    f"test_seed={exp.sampling_config.test_seed}",
+                ],
+            )
+            experiment_name = shortener(full_name, 4)
```
Comment (MischaPanch) [inline tianshou/highlevel/experiment.py]: I think instead the experiment needs a name. Why does one need to return a dict instead of a list of experiments?

File: examples/mujoco/mujoco_ppo_hl_multi.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python3
```
Comment (MischaPanch) [inline examples/mujoco/mujoco_ppo_hl_multi.py:1]: There should be a docstring explaining what the script does and how it's different from the other scripts. To work well with CLI tools, should probably be a docstring of the main function

**Review by MischaPanch — COMMENTED** [review]
Preliminary review, we can have a closer look together later,

One thing to do already is to separate saving plots from plotting, and/or give the user the possibility to configure how plots should be saved

File: examples/mujoco/mujoco_ppo_hl_multi.py (hunk 2 above)
Comment (maxhuettenrauch) [reply examples/mujoco/mujoco_ppo_hl_multi.py]: of course, forgot to comment in after testing

File: examples/mujoco/tools.py (hunk 3 above)
Comment (maxhuettenrauch) [reply examples/mujoco/tools.py]: score_threshold can default to None, then it is just inferred from the rewards achieved by the training runs. Regarding the name, maybe we can extract a default from the loaded experiment or from the exp_dir

File: examples/mujoco/tools.py (hunk 4 above)
Comment (maxhuettenrauch) [reply examples/mujoco/tools.py]: It's basically the number of times the trainer calls test_step(), which is usually once in reset and once after each epoch. We should find a better a way to tell the user what dimensions rliable expects the ndarrays to have (i.e., the arrays in score_dict should be 2D with each row representing a training run and env_steps should be 1D)

File: examples/mujoco/tools.py (hunk 5 above)
Comment (maxhuettenrauch) [reply examples/mujoco/tools.py]: rliable is actually not too heavyweight, I just kept it in there until we decide which way we go. I'll start with adding an extra `eval`

File: examples/mujoco/tools.py (hunk 6 above)
Comment (maxhuettenrauch) [reply examples/mujoco/tools.py]: The 0.25 makes makes the Q in IQM :) The reps parameter I need to check

File: examples/mujoco/tools.py (hunk 7 above)
Comment (maxhuettenrauch) [reply examples/mujoco/tools.py]: True, I just let the function return the ax objects and added a save_figure flag that defaults to None.

File: tianshou/highlevel/experiment.py (hunk 10 above)
Comment (maxhuettenrauch) [reply tianshou/highlevel/experiment.py]: Just wanted to avoid excessivley long dir names :D Yeah, let's discuss

File: tianshou/highlevel/experiment.py (hunk 11 above)
Comment (maxhuettenrauch) [reply tianshou/highlevel/experiment.py]: The experiment's path is composed of two parts. The persistence_dir in the experiment_config and a name given to run() of the Experiment. The name created here is just a convenience suggestion to have a (somewhat) meaningful default. It can obviosuly just be a list, too, and the user has to come up with their own names

File: examples/mujoco/tools.py (hunk 5 above)
Comment (MischaPanch) [reply examples/mujoco/tools.py]: I'll push the solution

File: tianshou/utils/logger/wandb.py [hunk 13]
```diff
@@ -156,3 +160,8 @@ def restore_data(self) -> tuple[int, int, int]:
         except KeyError:
             env_step = 0
         return epoch, env_step, gradient_step
+
+    def restore_logged_data(self, log_path: str) -> dict:
+        # TODO: add support for restoring logged data from W&B
+        assert self.tensorboard_logger is not None
```
Comment (MischaPanch) [inline tianshou/utils/logger/wandb.py]: replace by value error or remove optionality of the field

File: examples/atari/atari_wrapper.py [hunk 14]
```diff
@@ -340,7 +340,7 @@ def make_atari_env(
 
     :return: a tuple of (single env, training envs, test envs).
     """
-    env_factory = AtariEnvFactory(task, seed, frame_stack, scale=bool(scale))
+    env_factory = AtariEnvFactory(task, seed, seed + training_num, frame_stack, scale=bool(scale))
```
Comment (bordeauxred) [inline examples/atari/atari_wrapper.py:386]:
```suggestion
    env_factory = AtariEnvFactory(task, seed, seed + training_num + 1, frame_stack, scale=bool(scale))
```

Comment (bordeauxred) [reply examples/atari/atari_wrapper.py:386]: Add the  + 1 for consistency with test_seed property in config.py

File: examples/mujoco/mujoco_env.py [hunk 15]
```diff
@@ -29,7 +29,7 @@ def make_mujoco_env(task: str, seed: int, num_train_envs: int, num_test_envs: in
 
     :return: a tuple of (single env, training envs, test envs).
     """
-    envs = MujocoEnvFactory(task, seed, obs_norm=obs_norm).create_envs(
+    envs = MujocoEnvFactory(task, seed, seed + num_train_envs, obs_norm=obs_norm).create_envs(
```
Comment (bordeauxred) [inline examples/mujoco/mujoco_env.py:40]:
```suggestion
    envs = MujocoEnvFactory(task, seed, seed + num_train_envs + 1, obs_norm=obs_norm).create_envs(
```

File: tianshou/utils/logger/pandas_logger.py [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+            test_interval,
+            update_interval,
+            info_interval,
+            exclude_arrays,
+        )
+        self.log_path = log_dir
+        self.csv_name = os.path.join(self.log_path, "log.csv")
+        self.pkl_name = os.path.join(self.log_path, "log.pkl")
+        self.data: dict[str, list] = defaultdict(list)
+        self.last_save_step = -1
```
Comment (bordeauxred) [inline tianshou/utils/logger/pandas_logger.py]: Can we make some attributes private?

Comment (maxhuettenrauch) [reply tianshou/utils/logger/pandas_logger.py]: removed some of them. from the leftovers, which one should be private?

File: examples/mujoco/tools.py (hunk 3 above)
Comment (maxhuettenrauch) [reply examples/mujoco/tools.py]: Did a larger refactor to separate loading from preparing for rliable (new location is still not quite right though, as long as rliable is not in standard dependencies right now)

File: tianshou/highlevel/experiment.py (hunk 8 above)
Comment (maxhuettenrauch) [reply tianshou/highlevel/experiment.py]: removed

File: tianshou/utils/logger/wandb.py (hunk 13 above)
Comment (maxhuettenrauch) [reply tianshou/utils/logger/wandb.py]: done

File: tianshou/utils/logger/tensorboard.py [hunk 17]
```diff
@@ -132,14 +132,17 @@ def restore_data(self) -> tuple[int, int, int]:
...
+    ) -> dict[str, dict[str, VALID_LOG_VALS_TYPE | dict[str, VALID_LOG_VALS_TYPE]]]:
         ea = event_accumulator.EventAccumulator(log_path)
         ea.Reload()
 
         def add_to_dict(data_dict: dict[str, Any], keys: list[str], value: Any) -> None:
             current_dict = data_dict
-            for key in keys[:-1]:
-                current_dict = current_dict.setdefault(key, {})
+            for k in keys[:-1]:
+                current_dict = current_dict.setdefault(k, {})
```
Comment (bordeauxred) [inline tianshou/utils/logger/tensorboard.py]: How about using a defaultdict?

File: tianshou/utils/logger/tensorboard.py [hunk 18]
```diff
@@ -132,14 +132,17 @@ def restore_data(self) -> tuple[int, int, int]:
...
             current_dict[keys[-1]] = value
```
Comment (bordeauxred) [inline tianshou/utils/logger/tensorboard.py]: Since only the last key is assigned a value, I guess this is a fairly specialised add_to_dict. 
You might wish to reflect this in the naming.

File: examples/mujoco/mujoco_ppo_hl_multi.py (hunk 1 above)
Comment (maxhuettenrauch) [reply examples/mujoco/mujoco_ppo_hl_multi.py]: I added a very basic joblib launcher that allows for parallel experiments on a single machine. We can think about other launchers as well, but that's probably really out of scope.

File: examples/atari/atari_wrapper.py (hunk 14 above)
Comment (maxhuettenrauch) [reply examples/atari/atari_wrapper.py:386]: Actually the + 1 is not necessary. I also removed it from config.py

File: tianshou/utils/logger/pandas_logger.py (hunk 16 above)
Comment (maxhuettenrauch) [reply tianshou/utils/logger/pandas_logger.py]: removed the pandas logger from this PR

**Comment by maxhuettenrauch** [comment]
I had to use contextlib.suppress in order to make the docs build, but at least tests are passing now. I also linked our fork of rliable in the project dependencies as long as the original does not update the dependency on arch.

**Comment by MischaPanch** [comment]
> I also linked our fork of rliable in the project dependencies as long as the original does not update the dependency on arch.

Thanks, that works! Seems like rliable is no longer maintained, my PR bumping the arch version got no attention. Let's see for how long we don't need to touch it again. If it creates further problems, might be worth to move the functionality over to tianshou or somewhere else

**Comment by MischaPanch** [comment]
@maxhuettenrauch  as to the contextlib suppression - I don't think that's the best solution. Instead you could adjust the command in the CI, which is currently `poetry install --with dev` to `poetry install --with dev -E evaluation` (or something like that).

Note that you can run github actions locally with [this](https://github.com/nektos/act), maybe it makes sense to have a poe command for it

**Comment by maxhuettenrauch** [comment]
Thanks for the recommendation, unfortunately I haven't managed to get act to work (getting `::error::Not Found
` but no idea what is actually not found)

**Comment by MischaPanch** [comment]
Seems like this is ready, right? Gonna run a few tests tomorrow and then merge it

File: tianshou/utils/logger/tensorboard.py (hunk 18 above)
Comment (MischaPanch) [reply tianshou/utils/logger/tensorboard.py]: The logic is a bit complicated here, I renamed the function name and added more comments to explain it. Defaultdict wouldn't help here, unless there's something like an infinitely nested default dict ^^

**Comment by MischaPanch** [comment]
I couldn't make parallel execution work with joblib. It also might not make too much sense on a single machine, so we can consider using ray for parallelization instead. For now the limitations are documented, so I'd merge this.

In my last commits I improved the plots a bit (axis labels were cut off), added more docs, fixed some installation problems and did minor enhancements to the interfaces

**Review by MischaPanch — APPROVED** [review]
